lib/kbuildbarn/urls.go

```diff
@@ -0,0 +1,130 @@
+package kbuildbarn
+
+import (
+	"fmt"
+	"github.com/enfabrica/enkit/lib/multierror"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// ByteStreamUrl retieves the CAS id and bytes of an action based on the input url.
+// For example, bytestream://build.local.enfabrica.net:8000/blobs/a9a664559b4d29ecb70613fad33acfb287f2fa378178e131feaaebb5dafa231a/465
+// should return (a9a664559b4d29ecb70613fad33acfb287f2fa378178e131feaaebb5dafa231a, 465, nil)
+// which is a BuildBarnParams.Hash, BuildBarnParams.Size, error.Error
+func ByteStreamUrl(byteStream string) (string, string, error) {
+	u, err := url.Parse(byteStream)
+	if err != nil {
+		return "", "", err
+	}
+	splitUrl := strings.Split(u.Path, "/")
+	if len(splitUrl) != 4 {
+		return "", "", fmt.Errorf("ByteStreamUrl() bytestream url is not well formed %s", byteStream)
+	}
+	return splitUrl[2], splitUrl[3], nil
+}
+
+// BuildBarnParams are the parameter necessary to reverse proxy to a bb_browser instance.
+type BuildBarnParams struct {
+	FileName     string
+	Hash         string
+	Size         string
+	InvocationID string
+
+	// This is the base Url
+	BaseUrl string
+	Scheme  string
+
+	// These are the default buildbarn templates for their different types inside the CAS
+	FileTemplate      string
```

Comment:
nit: Suggest listing these four items in alphabetical order for ease of future maintenance. Likewise for the Default templates and in NewBuildBarmParams().

Reply:
Done